crates/nu-protocol/src/value/record.rs

```diff
@@ -0,0 +1,99 @@
+use crate::Value;
+
+use serde::{Deserialize, Serialize};
+
+#[derive(Debug, Clone, Default, Serialize, Deserialize)]
+pub struct Record {
+    pub cols: Vec<String>,
+    pub vals: Vec<Value>,
+}
+
+impl Record {
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    pub fn with_capacity(capacity: usize) -> Self {
+        Self {
+            cols: Vec::with_capacity(capacity),
+            vals: Vec::with_capacity(capacity),
+        }
+    }
+
+    pub fn iter(&self) -> Iter {
+        self.into_iter()
+    }
+
+    pub fn iter_mut(&mut self) -> IterMut {
+        self.into_iter()
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.cols.is_empty() || self.vals.is_empty()
+    }
+
+    pub fn len(&self) -> usize {
+        usize::min(self.cols.len(), self.vals.len())
+    }
+
+    pub fn push(&mut self, col: impl Into<String>, val: Value) {
+        self.cols.push(col.into());
+        self.vals.push(val);
+    }
+}
+
+impl FromIterator<(String, Value)> for Record {
+    fn from_iter<T: IntoIterator<Item = (String, Value)>>(iter: T) -> Self {
+        let (cols, vals) = iter.into_iter().unzip();
+        Self { cols, vals }
+    }
+}
+
+pub type IntoIter = std::iter::Zip<std::vec::IntoIter<String>, std::vec::IntoIter<Value>>;
+
+impl IntoIterator for Record {
+    type Item = (String, Value);
+
+    type IntoIter = IntoIter;
+
+    fn into_iter(self) -> Self::IntoIter {
+        self.cols.into_iter().zip(self.vals)
+    }
+}
+
+pub type Iter<'a> = std::iter::Zip<std::slice::Iter<'a, String>, std::slice::Iter<'a, Value>>;
+
+impl<'a> IntoIterator for &'a Record {
+    type Item = (&'a String, &'a Value);
+
+    type IntoIter = Iter<'a>;
+
+    fn into_iter(self) -> Self::IntoIter {
+        self.cols.iter().zip(&self.vals)
+    }
+}
+
+pub type IterMut<'a> = std::iter::Zip<std::slice::Iter<'a, String>, std::slice::IterMut<'a, Value>>;
+
+impl<'a> IntoIterator for &'a mut Record {
+    type Item = (&'a String, &'a mut Value);
```

Comment:
I left the key/column immutable here. Should this be mutable?

Reply:
Sounds good to have immutable key references should we opt to replace the underlying data to a map type that explicitly requires this. And the existing `Record` assumptions would also be broken if this were mutable.

Good choice